swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -139,85 +149,95 @@ class Adapter {
+    let deviceInfoStr = String(data: deviceInfo, encoding: .utf8) ?? "{}"
+    let logDir = SharedAccess.connlibLogFolderURL?.path ?? "/tmp/firezone"
 
-    // Cancel network monitor
-    networkMonitor?.cancel()
-  }
+    // Create the session
+    let session: Session
+    do {
+      session = try Session.newApple(
+        apiUrl: apiURL,
+        token: token.description,
+        deviceId: deviceId,
+        accountSlug: accountSlug,
+        deviceName: deviceName,
+        osVersion: osVersion,
+        logDir: logDir,
+        logFilter: logFilter,
+        deviceInfo: deviceInfoStr,
+        isInternetResourceActive: internetResourceEnabled
+      )
+    } catch {
+      throw AdapterError.connlibConnectError(String(describing: error))
+    }
 
-  /// Start the tunnel.
-  func start() throws {
-    Log.log("Adapter.start")
+    // Create channels - following Rust pattern with separate sender/receiver
+    let (commandSender, commandReceiver): (Sender<SessionCommand>, Receiver<SessionCommand>) =
+      Channel.create()
+    self.commandSender = commandSender
+
+    let (eventSender, eventReceiver): (Sender<Event>, Receiver<Event>) = Channel.create()
+    self.eventReceiver = eventReceiver
+
+    // Start event loop - owns session, receives commands, sends events
+    Task { [weak self] in
+      defer {
+        // ALWAYS cleanup, even if event loop crashes
+        self?.commandSender = nil
+        self?.eventReceiver = nil
+        Log.log("Adapter: Event loop finished, session dropped")
+      }
 
-    guard session == nil else {
-      throw AdapterError.invalidSession(session)
+      await runSessionEventLoop(
+        session: session,
+        commandReceiver: commandReceiver,
+        eventSender: eventSender
+      )
     }
 
-    callbackHandler.delegate = self
+    // Start event consumer - consumes events from receiver (Rust pattern: receiver outside)
+    Task { [weak self] in
+      guard let eventReceiver = self?.eventReceiver else { return }
 
-    Log.log("Adapter.start: Starting connlib")
-    do {
-      let jsonEncoder = JSONEncoder()
-      jsonEncoder.keyEncodingStrategy = .convertToSnakeCase
-
-      // Grab a session pointer
-      session = try WrappedSession.connect(
-        apiURL,
-        "\(token)",
-        "\(id)",
-        accountSlug,
-        DeviceMetadata.getDeviceName(),
-        DeviceMetadata.getOSVersion(),
-        connlibLogFolderPath,
-        logFilter,
-        internetResourceEnabled,
-        callbackHandler,
-        String(data: jsonEncoder.encode(DeviceMetadata.deviceInfo()), encoding: .utf8)!
-      )
-    } catch let error {
-      // `toString` needed to deep copy the string and avoid a possible dangling pointer
-      let msg = (error as? RustString)?.toString() ?? "Unknown error"
-      throw AdapterError.connlibConnectError(msg)
+      for await event in eventReceiver.stream {
+        await self?.handleEvent(event)
```

Comment:
Can `handleEvent` fail? If yes, we should wrap this in a try catch and log an error but continue operating.

Reply:
Given that `self` is a weak-reference here, I think we should use a `guard` that breaks out of the loop if `self` ends up being nil. The guard needs to be inside the loop though to make sure we check it on each iteration.

Reply:
> Can `handleEvent` fail? If yes, we should wrap this in a try catch and log an error but continue operating.

yes, it can fail, only if the JSON decoding fails - the other case is wrapped in try/catch block.  I guess if we address #9548 we might be able to make it infallible? but until then, I'll wrap it in try/catch

> Given that self is a weak-reference here, I think we should use a guard that breaks out of the loop if self ends up being nil. The guard needs to be inside the loop though to make sure we check it on each iteration.

good catch.